Add new Rabbit Air integration

al-s opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add new integration for Rabbit Air air purifiers.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/21560

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

al-s committed:
Add new Rabbit Air integration

homeassistant commented:
Hello @al-s,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`a3ec34d0533836cbc129cea93741b840614b8bc9`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;66130&#x2F;commits&#x2F;a3ec34d0533836cbc129cea93741b840614b8bc9) - This commit has something that looks like an email address (alexs@pushok.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hi @al-s, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#66130).

Once you do that we will be able to review and accept this pull request.

Thanks!

DCSBL reviewed (commented):
Nice to see this going into core!
I've added some comments that maybe require some attention.

Also, there is a weird `py.typed` file in your PR. If you want to enable strict type checks, you can add the files to `/.strict-typing`

al-s committed:
Remove py.typed file

It is not needed and was just accidentally added to the commit.

al-s committed:
Enable strict type checking for rabbitair component

Keeping the code fully type hinted is a good idea.

al-s committed:
Add missing type annotations

al-s committed:
Remove translation file

al-s committed:
Prevent data to be added to hass.data if refresh fails

al-s committed:
Reload the config entry when the options change

al-s committed:
Add missing type parameters for generics

al-s committed:
Avoid using assert in production code

al-s committed:
Move zeroconf to optional dependencies

al-s commented on homeassistant/components/rabbitair/__init__.py:
```diff
@@ -0,0 +1,76 @@
+    zeroconf_instance = await zeroconf.async_get_async_instance(hass)
+    device: Client = UdpClient(host, token, zeroconf=zeroconf_instance)
+
+    scan_interval: int = entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
```
Thanks for the tip! The changes were applied directly from options flow, but it is better to reload the config entry. In the future, more options will be added for which it will be difficult to apply changes on the fly, without reloading.

al-s commented on homeassistant/components/rabbitair/__init__.py:
```diff
@@ -0,0 +1,76 @@
+        KEY_DEVICE: device,
+    }
+
+    await coordinator.async_config_entry_first_refresh()
```
Thanks! Fixed.

al-s commented on homeassistant/components/rabbitair/entity.py:
```diff
@@ -0,0 +1,69 @@
+        self._entry = entry
+        # Use assert to suppress the type warning. We always set a unique ID in
+        # the config flow, so it should never be None.
+        assert entry.unique_id is not None
```
Replaced it with `# type: ignore` annotation, just to avoid using `assert`. Although, in my opinion, this solution is worse, because it hides possible errors in our code and, therefore, is less safe.

al-s commented on homeassistant/components/rabbitair/manifest.json:
```diff
@@ -0,0 +1,21 @@
+  "homekit": {},
+  "dependencies": [
+    "zeroconf"
+  ],
```
Not sure about this. The [docs](https://developers.home-assistant.io/docs/network_discovery/#mdnszeroconf) say that we should add `zeroconf` to `dependencies` if we use `async_get_async_instance`. Perhaps with the `zeroconf` section, this dependency is added implicitly, but I have not found this documented in any way. However, I have noticed that most integrations specify it in `after_dependencies` (i.e. as an optional dependency), and this is more appropriate since having mDNS functionality is not an absolute requirement.

al-s commented on homeassistant/components/rabbitair/translations/en.json:
```diff
@@ -0,0 +1,31 @@
+            }
+        }
+    }
+}
```
Thanks! Fixed.

frenck commented on homeassistant/components/rabbitair/__init__.py:
```diff
@@ -0,0 +1,83 @@
+    host: str = entry.data[CONF_HOST]
+    token: str = entry.data[CONF_ACCESS_TOKEN]
+
+    _LOGGER.debug("Initializing with host %s, token %s", host, token[:5])
```
The integration is already logged by Home Assistant and we should never log tokens :)

```suggestion
```

frenck commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+            step_id="init",
+            data_schema=vol.Schema(
+                {
+                    vol.Optional(
+                        CONF_SCAN_INTERVAL,
+                        default=scan_interval,
+                    ): vol.All(vol.Coerce(int), vol.Range(min=2)),
```
Scan interval should not be part of integration configuration.
If someone wants a custom polling interval or schema, an automation that calls the `homeassistant.entity_update` service instead.

This mean this option flow can be fully removed, as this was the only option.

al-s replied:
Ok, removed it.

frenck commented on homeassistant/components/rabbitair/entity.py:
```diff
@@ -0,0 +1,75 @@
+        # Wait for the changes to be applied and for the device status to be
+        # updated. Two seconds should be sufficient, since the internal cycle
+        # of the device runs at one-second intervals.
+        await asyncio.sleep(2)
```
We should change the cooldown and set immediate refresh on the coordinator instead.

al-s replied:
Ok, the delay is now implemented with the update coordinator.

frenck commented on homeassistant/components/rabbitair/entity.py:
```diff
@@ -0,0 +1,75 @@
+        # causes other parameters to change as well. By getting updated status
+        # immediately we provide a better user experience, especially if the
+        # default polling interval is set too long.
+        await self.coordinator.async_refresh()
```
Once the cooldown/immediate update has been adjusted on the coordinator, you can adjust this:

```suggestion
        await self.coordinator.async_request_refresh()
```

al-s replied:
This turned out to be a little more complicated. When I just set cooldown=2 and immediate=false, there was a weird effect in the user interface: after changing the state of the controls, they could go back to their previous state. It turned out that the effect is composed of two parts.
1. There was the possibility of polling the device before 2 seconds had elapsed and receiving a still unchanged state. This can happen either because of periodic polling (it happens immediately at the scheduled time, and cancels the cooldown delay), or because several requests were sent during the cooldown period (since the delay time does not start counting again).
2. There seems to be a 2-second timeout somewhere in the user interface. This is especially noticeable on the power switch. If the state reported by the entity has not changed within this time, the switch returns to its previous state. But then, almost immediately, a new state comes from the device, and the switch jumps back again.

I made some changes to avoid this effect. Could you take a look? Maybe there is some other, more correct, solution?

frenck commented on homeassistant/components/rabbitair/entity.py:
```diff
@@ -0,0 +1,75 @@
+    @property
+    def name(self) -> str | None:
+        """Return the name of this entity, if any."""
+        return self._entry.title or "RabbitAir"
```
When is entry title empty?

frenck replied:
Additionally, this could be set as an entity attribute (`self._attr_name = "X" in the constructor).

al-s replied:
The title can be empty if the user has explicitly set it to empty string.

Replaced the property with an attribute.

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,132 @@
+}
+
+
+class RabbitAirFanEntity(RabbitAirBaseEntity, FanEntity):
+    """Fan control functions of the Rabbit Air air purifier."""
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return SUPPORT_PRESET_MODE | SUPPORT_SET_SPEED
```
```suggestion
class RabbitAirFanEntity(RabbitAirBaseEntity, FanEntity):
    """Fan control functions of the Rabbit Air air purifier."""
    
    _attr_supported_features = SUPPORT_PRESET_MODE | SUPPORT_SET_SPEED

```

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,132 @@
+        speed = self.coordinator.data.speed
+        return (
+            None
+            if speed is None
+            else 0
+            if speed is Speed.SuperSilent
+            else ordered_list_item_to_percentage(SPEED_LIST, speed)
```
We should not use ternaries when they split across multiple lines, it makes code hard to read.

al-s replied:
Fixed.

frenck commented on homeassistant/components/rabbitair/__init__.py:
```diff
@@ -0,0 +1,83 @@
+
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data[DOMAIN][entry.entry_id] = {
```
We could make a NamedTuple to hold the coordinator and device. It would be easier to use and easier to maintain typing.

al-s replied:
Thanks for the tip! I made that change. But then, after rewriting the code related to the update coordinator, it became sufficient to pass only the coordinator itself, so I removed the NamedTuple.

frenck commented on homeassistant/components/rabbitair/entity.py:
```diff
@@ -0,0 +1,75 @@
+import logging
+from typing import Any
+
+from rabbitair import Client, Model, State
```
Client, Model & State are not re-exported from the package root.

```suggestion
from rabbitair.client import Client
from rabbitair.response import Model, State
```

al-s replied:
Fixed the library to properly re-export these imports.

frenck commented on homeassistant/components/rabbitair/__init__.py:
```diff
@@ -0,0 +1,83 @@
+from datetime import timedelta
+import logging
+
+from rabbitair import Client, State, UdpClient
```
These imports are not re-exported by the package. Import them from the appropriate modules instead.

al-s replied:
Fixed the library to properly re-export these imports.

frenck commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+import logging
+from typing import Any
+
+from rabbitair import UdpClient
```
This import is not re-exported by the package. Import it from the appropriate modules instead.

al-s replied:
Fixed the library to properly re-export these imports.

frenck commented on homeassistant/components/rabbitair/entity.py:
```diff
@@ -0,0 +1,75 @@
+        self._entry = entry
+        self._attr_unique_id = entry.unique_id
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, entry.unique_id)},  # type: ignore[arg-type]
```
Why are we ignoring type here? Can we avoid doing that?

al-s replied:
This is because, according to the type information, `entry.unique_id` can be `None`, while `identifiers` can't. But we always fill in the `unique_id` when we create an entry, so it should not be `None` "by design". Perhaps we could use assert here? That would also remove the warning, and better show our intentions.

frenck replied:
Made a suggestion in the PR to avoid needing to do this (by storing the MAC in the entry data)

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,132 @@
+from .const import DOMAIN, KEY_COORDINATOR, KEY_DEVICE
+from .entity import RabbitAirBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
```
Seems to be unused?

al-s replied:
Yes, removed.

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,132 @@
+        """Turn the fan off."""
+        await self._set_state(power=False)
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if device is on."""
+        return self.coordinator.data.power is True
```
`power` can be `None` according to typing, why not return that?

```suggestion
    @property
    def is_on(self) -> bool | None:
        """Return true if device is on."""
        return self.coordinator.data.power
```

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,132 @@
+    def preset_modes(self) -> list[str] | None:
+        """Return a list of available preset modes."""
+        if self._is_model(Model.MinusA2):
+            return list(PRESET_MODES.keys())
```
```suggestion
            return list(PRESET_MODES)
```

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,132 @@
+        return next(k for k, v in PRESET_MODES.items() if v == mode)
+
+    @property
+    def preset_modes(self) -> list[str] | None:
```
We can also set this once in the constructor (`self._attr_preset_modes = X`)

al-s replied:
Fixed.

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,132 @@
+        )
+
+    @property
+    def speed_count(self) -> int:
```
We can also set this one in the constructor (once on init), `self._attr_speed_count = X`

al-s replied:
Fixed.

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,132 @@
+        return None
+
+
+async def async_setup_entry(
```
Small detail, but can we move this up in the file?

al-s replied:
Fixed.

frenck commented on homeassistant/components/rabbitair/manifest.json:
```diff
@@ -0,0 +1,21 @@
+  "requirements": [
+    "python-rabbitair==0.0.6"
+  ],
+  "ssdp": [],
```
```suggestion
```

frenck commented on homeassistant/components/rabbitair/manifest.json:
```diff
@@ -0,0 +1,21 @@
+  "zeroconf": [
+    "_rabbitair._udp.local."
+  ],
+  "homekit": {},
```
```suggestion
```

frenck commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+            errors["base"] = "unknown"
+        else:
+            await self.async_set_unique_id(dr.format_mac(info["mac"]))
+            self._abort_if_unique_id_configured()
```
Maybe update the existing entry in that case?

al-s replied:
Yes, that's a good idea.

frenck commented on homeassistant/components/rabbitair/manifest.json:
```diff
@@ -0,0 +1,21 @@
+    "python-rabbitair==0.0.6"
+  ],
+  "ssdp": [],
+  "zeroconf": [
```
I don't see the configuration flow step for this implemented?

al-s replied:
Added a step for zeroconf.

I have a question about this. I tried to update `CONFIG_HOST` with the current IP, using `_abort_if_unique_id_configured(updates={CONFIG_HOST: discovery_info.host})` on zeroconf events. But if such a change took place at the moment of starting `hass`, an error would appear in the logs:
```
ERROR (MainThread) [homeassistant.config_entries] Error unloading entry RabbitAir-XXXXXXXXXXXX for fan
Traceback (most recent call last):
  File "core/homeassistant/config_entries.py", line 467, in async_unload
    result = await component.async_unload_entry(hass, self)
  File "core/homeassistant/components/fan/__init__.py", line 195, in async_unload_entry
    return await component.async_unload_entry(entry)
  File "core/homeassistant/helpers/entity_component.py", line 172, in async_unload_entry
    raise ValueError("Config entry was never loaded!")
ValueError: Config entry was never loaded!
```
However, it did not affect anything, after that everything worked.

If the configuration was updated from the same handler when `hass` was already running, no errors occurred.

What could be the reason? Maybe I'm doing something wrong?

I have now removed this update as it's not really needed that much. The rabbitair library can resolve the mDNS names itself if a zeroconf object was provided during initialization.

frenck commented on homeassistant/components/rabbitair/manifest.json:
```diff
@@ -0,0 +1,21 @@
+  ],
+  "homekit": {},
+  "after_dependencies": [
+    "zeroconf"
```
How is `zeroconf` used that it is needed as an after dependency?

al-s replied:
An instance of AsyncZeroconf is passed to the library so that it can resolve mDNS names. And, according to the [docs](https://developers.home-assistant.io/docs/network_discovery#mdnszeroconf), in this case `zeroconf` should be added to `dependencies` in the integration's `manifest.json`. But since mDNS functionality is optional, `after_dependencies` is used instead of `dependencies` (and I see the same in other integrations).

It should be noted that the current firmware does not support zeroconf discovery, but does support mDNS. However, this should not be a problem. The next update will add full support for zeroconf.

frenck reviewed (changes requested):
Hi there @al-s,

Made a first review pass on your PR and left some comments. Could you take a look? 

Thanks! 👍 

../Frenck

al-s committed:
Remove unnecessary logging

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

al-s committed:
Remove unused keys from the manifest

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

al-s committed:
Replace property with attr

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

al-s committed:
Allow to return None for power

The type of the is_on property now allows this.

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

al-s committed:
Remove unnecessary method call

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

al-s committed:
Update the python library

The new version properly re-exports names from the package root.

al-s committed:
Remove options flow

Scan interval should not be part of integration configuration. This was
the only option, so the options flow can be fully removed.

al-s committed:
Replace properties with attrs

al-s committed:
Remove multiline ternary operator

al-s committed:
Use NamedTuple for hass.data

al-s committed:
Remove unused logger variable

al-s committed:
Move async_setup_entry up in the file

al-s committed:
Adjust debouncer settings to use request_refresh

al-s committed:
Prevent status updates during the cooldown period

al-s committed:
Move device polling code to the update coordinator

al-s committed:
Fix the problem with the switch jumping back and forth

The UI seems to have a timeout of 2 seconds somewhere, which is just a
little bit less than what we normally need to get an updated state. So
the power switch would jump to its previous state and then immediately
return to the new state.

al-s committed:
Update the python library

The new version fixes errors when multiple requests are executed
simultaneously.

al-s committed:
Fix incorrect check for pending call in debouncer

This caused the polling to stop.

al-s committed:
Fix tests

* Update .coveragerc to exclude new file.
* Remove test for Options Flow.

al-s committed:
Update the existing entry when device access details change

al-s committed:
Add Zeroconf discovery step

al-s commented:
Rebased on top of dev branch to the the latest changes and resolve conflicts.

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

al-s committed:
Fix tests

The ZeroconfServiceInfo constructor now requires one more argument.

al-s committed:
Fix typing for CoordinatorEntity

al-s committed:
Fix signature of async_turn_on

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

al-s committed:
Fix depreciation warnings

al-s committed:
Fix manifest formatting

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

al-s committed:
Fix warning about debouncer typing

relates to 5ae5ae5392729b4c94a8004bd02e147d60227341

al-s committed:
Wait for config entry platform forwards

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

frenck commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,125 @@
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST, default=host): str,
```
To make you life easier (in terms of typing) down the line, you might want to consider putting the MAC address in the config entry data as well.
Right now, you seem to rely the config entry unique ID in those cases, that works, but since unique ID is optional typed, that makes it a tiny bit harder.

al-s replied:
Implemented. The MAC address is now stored in the config entry data.

frenck commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,125 @@
+        """Handle zeroconf discovery."""
+        mac = dr.format_mac(discovery_info.properties["id"])
+        await self.async_set_unique_id(mac)
+        self._abort_if_unique_id_configured()
```
This method accepts a `updates` parameter, which can update the existing entry. You could use this to update the CONF_HOST based on discovered data.
It helps when the IP address changed on the network of the user, causing it to be picked up automatically again.

al-s replied:
Thanks for the tip!
Right now the mDNS name is used to access the device. It allows not only to find the device in the network, but also to distinguish between devices if the user has more than one, as the user will need to enter the correct access key for the specific unit. The mDNS name together with the access key are displayed in the mobile app on the device page.
The user has the option to enter the IP address of the device instead of the hostname, for example, if mDNS does not work because the device is on a different network. However, it is likely that Zeroconf discovery will not work in this case either.

frenck commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,125 @@
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> FlowResult:
+        """Handle zeroconf discovery."""
+        mac = dr.format_mac(discovery_info.properties["id"])
```
Same as above, there is no need to mac format the unique ID for a config entry.

al-s replied:
Removed the format_mac call. But formatting is still required, because different forms of MAC-address representation are possible. In this case, the string received from Zeroconf does not match the string returned by the device API.

emontnemery replied:
I discussed this with Frenck, the documentation is indeed correct.
Please revert back to calling `format_mac`

frenck commented on homeassistant/components/rabbitair/coordinator.py:
```diff
@@ -0,0 +1,77 @@
+            request_refresh_debouncer=RabbitAirDebouncer(hass),
+        )
+
+    async def _async_update_data(self) -> State:
+        try:
+            return await self.device.get_state()
+        except asyncio.TimeoutError:
+            raise
+        except Exception as err:
+            raise UpdateFailed from err
```
These exceptions are all already handled by the DateUpdateCoordinator and don't need to be catched here.

al-s replied:
Removed this extra handling. It allowed to make the log a bit cleaner. The default handler logged such exceptions with the "error" level and a long backtrace. In a situation where the device just does not respond, this is not necessary.

frenck commented on homeassistant/components/rabbitair/entity.py:
```diff
@@ -0,0 +1,62 @@
+        """Initialize the entity."""
+        super().__init__(coordinator)
+        self._entry = entry
+        self._attr_name = entry.title or "Rabbit Air"
```
This can now be removed, it will automatically fallback to the device name when not set.

al-s replied:
Ok, removed.

frenck commented on homeassistant/components/rabbitair/entity.py:
```diff
@@ -0,0 +1,62 @@
+    ) -> None:
+        """Initialize the entity."""
+        super().__init__(coordinator)
+        self._entry = entry
```
The current config entry is already available in the coordinator: `self.coordinator.config_entry`, we don't need to store it here.

al-s replied:
Deleted that line. Looks like it was some code that was accidentally left behind and not even used.

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,148 @@
+
+        self._attr_speed_count = len(SPEED_LIST)
+
+        self._update_state()
```
We should not be updating the state here. The coordinator should have the latest state and the entity would reflect that.
The coordinator will signal an update state on the entity when it refreshed/fetches data.

al-s replied:
Perhaps the name is poorly chosen and misleading. This is just an internal helper method that transforms the raw data received from the device API and fills in the corresponding class fields that store the current state of the device.

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,148 @@
+        self._power = False
+        self.async_write_ha_state()
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return true if device is on."""
+        return self._power
```
This property can be removed, and replace `self._power` with `self._attr_is_on` that is already available.

al-s replied:
Unfortunately, that's not the case. The parent class FanEntity overrides this method so that it does not use the value of the _attr_is_on field. But we can override it again to get the previous behavior back.

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,148 @@
+        if percentage is not None:
+            speed_value = percentage_to_ordered_list_item(SPEED_LIST, percentage)
+        await self._set_state(power=True, mode=mode_value, speed=speed_value)
+        self._power = True
```
`self._attr_is_on = True`

al-s replied:
_power replaced with _attr_is_on

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,148 @@
+            self._attr_percentage = percentage
+        if preset_mode is not None:
+            self._attr_preset_mode = preset_mode
+        self.async_write_ha_state()
```
Shouldn't it schedule an update with the coordinator instead?

al-s replied:
If I understand correctly, the suggestion is to request an update with `await self.coordinator.async_request_refresh()`, and update the state with real data received from the device. Yes, we can do that. But then the UI will look slow or unresponsive. The reason is that we can't get the update quick enough. It makes no sense to poll the device less than 2 seconds after sending a state change command, otherwise the state may not have changed yet.

For example, if the user turns off the device or changes the speed, the controls in the UI will show these changes with a 2-second delay. However, the advantage is that it will be the real state, not what we assumed it should be.

Also, as far as I remember, there were problems with the previous UI: if we didn't update the state fast enough, the controls jumped from the new value to the old value and back again. But it seems that now, with the new UI, there is no such effect.

In contrast, the current variant will display changes immediately, which gives the feeling that the interface is responsive to user actions and provides good feedback. However, these changes do not reflect the actual state of the device, but only our assumptions about it. The real state will be fetched in a few seconds, on the next polling cycle.

Not sure which of these options is better.

frenck commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,148 @@
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the fan off."""
+        await self._set_state(power=False)
+        self._power = False
+        self.async_write_ha_state()
```
`self._attr_is_on = False`
Shouldn't it schedule an update with the coordinator instead?

al-s replied:
_power replaced with _attr_is_on

al-s replied:
About updating the state, answered above.

al-s committed:
Apply some of the suggested changes

* Do not put the MAC address in the title. Use a fixed title instead.
* Do not format the MAC to use as a unique ID.
* Do not catch exceptions in _async_update_data().
* Remove unused _entry field in the base entity class.
* Use the standard attribute self._attr_is_on to keep the power state.

al-s committed:
Store the MAC in the config entry data

michaelkkehoe commented:
Bump :)
Would love to see this merged

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

michaelkkehoe commented:
@al-s Bump. Is this still being worked on?

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

edenhaus reviewed (dismissed):
Thanks @al-s for your PR :)

There are still some conversation open from the last reviews. Can you please address them and close it afterwards.

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

al-s committed:
Change the order of except clauses

OSError is an ancestor class of TimeoutError, so TimeoutError should be
handled first

al-s committed:
Fix depreciation warnings

frenck commented:
@al-s I've seen you taking action on the last comment by @edenhaus, but.. conversations have not been marked as resolved and this PR is currently still in draft. Are you still working on it?

../Frenck

al-s commented:
@frenck @edenhaus Thanks for your attention to this PR! It's been a while since I wrote this code and I wanted to recheck that everything still works fine, but had no time for it. I'll try to do it in the next few days.

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

al-s committed:
Fix tests

The ZeroconfServiceInfo constructor arguments have changed.

al-s commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,125 @@
+        raise CannotConnect from err
+
+    # Return info to store in the config entry.
+    return {"title": f"RabbitAir-{info.mac.replace(':', '')}", "mac": info.mac}
```
Ok, the fixed title "Rabbit Air" will be used.

al-s commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,125 @@
+                _LOGGER.debug("Unexpected exception: %s", err)
+                errors["base"] = "unknown"
+            else:
+                await self.async_set_unique_id(dr.format_mac(info["mac"]))
```
Hmm, the [documentation](https://developers.home-assistant.io/docs/config_entries_config_flow_handler#example-acceptable-sources-for-a-unique-id) says to use format_mac in this case. Ok, removed MAC formatting.

al-s replied:
Thanks! I see you've already reverted the changes back.

neumachen commented:
Is there anything I can contribute/review to this to speed this integration's merge faster? I only have RabbitAir purifiers in my house, and this would help me. :D

edenhaus commented:
The PR is still in draft and, therefore, considered in progress and not ready for another review.
The author can mark it as ready for review when he thinks it is ready.

> Is there anything I can contribute/review to this to speed this integration's merge faster? I only have RabbitAir purifiers in my house, and this would help me. :D

Yes sure. You can review other PR's to bring the number of PR's down :) Less open PR's means we can come back to this one quicker :)

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

al-s committed:
Fix DeviceInfo import

al-s committed:
Rename the method to make it clearer what it does

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

emontnemery commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_HOST, CONF_MAC
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
```
```suggestion
from homeassistant.exceptions import HomeAssistantError
from homeassistant.helpers import device_registry as dr
```

emontnemery commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+                errors["base"] = "unknown"
+            else:
+                user_input[CONF_MAC] = info["mac"]
+                await self.async_set_unique_id(info["mac"])
```
```suggestion
                await self.async_set_unique_id(dr.format_mac(info["mac"]))
```

emontnemery commented on homeassistant/components/rabbitair/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> FlowResult:
+        """Handle zeroconf discovery."""
+        mac = discovery_info.properties["id"]
+        mac = ":".join(mac[i : i + 2] for i in range(0, 12, 2))
```
```suggestion
        mac = dr.format_mac(discovery_info.properties["id"])
```

emontnemery committed:
Apply suggestions from code review

emontnemery commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,148 @@
+    @property
+    def is_on(self) -> bool | None:
+        """Return true if device is on."""
+        return self._attr_is_on
```
I think @frenck didn't realize `FanEntity` ignores `_attr_is_on`. It's somewhat confusing to use `_attr_is_on` here and then also overriding the property of the base class.

Furthermore, the base class implements `is_on` like this:
```py
    @property
    def is_on(self) -> bool | None:
        """Return true if the entity is on."""
        return (
            self.percentage is not None and self.percentage > 0
        ) or self.preset_mode is not None
```

The custom implementation in this class will allow the fan to report a fan speed which is not None or 0 or a preset mode which is not None, yet the state will be `off`.
This seems illogical, why doesn't the `is_on` from the base class suffice?

@bdraco could you comment?

bdraco replied:
is_on should probably be marked as @final in the base fan entity

emontnemery replied:
OK, so if the API has a separate flag for if the fan is on, that should be used to conditionally set the percentage and preset_mode to 0 / None.

emontnemery replied:
> is_on should probably be marked as https://github.com/Final in the base fan entity

That's easier said than done, it seems about 2/3 of the fan entities override `is_on`.

al-s replied:
Changed the logic to use `is_on` from the base class. Now power off will cause the percentage and preset_mode to be set to None. Setting the percentage to 0 will cause the power to be turned off. Setting the percentage > 0 or preset_mode to some value will cause the unit to turn on.

emontnemery commented on homeassistant/components/rabbitair/fan.py:
```diff
@@ -0,0 +1,148 @@
+        if data.speed is None:
+            self._attr_percentage = None
+        elif data.speed is Speed.SuperSilent:
+            self._attr_percentage = 0
```
0 means the fan is off according to the [documentation](https://developers.home-assistant.io/docs/core/entity/fan). Maybe use 1 instead if this is a very low speed?

@bdraco WDYT?

bdraco replied:
1 or whatever super silent is as a percentages of maximum speed makes sense to me

al-s replied:
Will use 1% for Super Silent mode.

emontnemery reviewed (changes requested):
Please update the fan logic as per previous comments

al-s committed:
Merge remote-tracking branch 'upstream/dev' into rabbitair

al-s committed:
Fix tests

al-s committed:
Change speed/mode logic to use is_on from the base class

al-s committed:
A zero value is more appropriate than None

since None means "unknown", but we actually know that the speed is zero
when the power is off.

emontnemery reviewed (approved):
LGTM, thanks @al-s 👍

emontnemery committed:
Merge branch 'dev' into rabbitair
